### PR TITLE
[Pollen robotics] pollen vision blog post

### PR DESCRIPTION
**Pull request by apirrone**

### CODE REVIEW

File: pollen-vision.md [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+- OWL-Vit does not detect everything every time and can be inconsistent. We are looking for a better option.
+- There is no temporal or spatial consistency so far. All is recomputed every frame
+    - We are currently working on integrating a point tracking solution to enhance the consistency of the detections
+- Grasping technique (only front grasp for now) was not the focus of this work. We will be working on different approaches to enhance the grasping capabilities in terms of perception (6D detection) and grasping pose generation.
+- Overall speed could be improved
+
+### Try pollen-vision
+
+Wanna try pollen-vision? Check out the notebooks on Google Colab here https://drive.google.com/drive/folders/1Xx42Pk4exkS95iyD-5arHIYQLXyRWTXw.
```
Comment (osanseviero) [inline pollen-vision.md:113]: I suggest to host notebooks in GitHub or Hugging Face, as otherwise it's easy for these links to break

**Review by osanseviero — COMMENTED** [review]
Very cool :fire:

File: pollen-vision.md (hunk 1 above)
Comment (apirrone) [reply pollen-vision.md:113]: The notebooks are already hosted on our github repo here https://github.com/pollen-robotics/pollen-vision/tree/develop/examples/vision_models_examples

File: pollen-vision.md [hunk 2]
```diff
@@ -0,0 +1,115 @@
+---
```
[the remark this answers is not in the document]
Comment (apirrone) [reply pollen-vision.md:1]: Done !

File: pollen-vision.md [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+title: "Pollen-Vision: Unified interface for Zero-Shot vision models in robotics" 
+thumbnail: /blog/assets/pollen-vision/thumbnail.jpg
+authors:
+- user: apirrone
```
[the remark this answers is not in the document]
Comment (apirrone) [reply pollen-vision.md:5]: Done !

File: pollen-vision.md [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+- user: simheo
+- user: PierreRouanet
+- user: revellsi
+---
+
+# Pollen-Vision: Unified interface for Zero-Shot vision models in robotics
+
+<!-- TODO intro mp4 -->
+
+At Pollen Robotics, we're dedicated to simplifying robotics, making it more accessible and practical for everyday applications. For this reason, we have created [Reachy](https://www.youtube.com/watch?v=iSL39WFxCLE), an open source humanoid robot designed with manipulation in the real world in mind (80+ units sold around the world). It can be programmed through a simple Python [SDK](https://docs.pollen-robotics.com/docs/installation/use-sdk/) and [teleoperated via a VR headset](https://www.youtube.com/watch?v=lRWna6mEMv0).
```
[the remark this answers is not in the document]
Comment (apirrone) [reply pollen-vision.md]: What do you mean by "a banner sayint that this is a guest blog post" ?

File: pollen-vision.md (hunk 1 above)
Comment (osanseviero) [reply pollen-vision.md:117]: But the link is still to Google Drive. Why not link to https://github.com/pollen-robotics/pollen-vision/tree/develop/examples/vision_models_examples?

File: pollen-vision.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (osanseviero) [reply pollen-vision.md]: Maybe something like

```
> [!NOTE]  
> This is a guest blog post by the Pollen Robotics team, a startup focused on ...
```

or something on those lines

[the remark this answers is not in the document]
Comment (apirrone) [reply pollen-vision.md]: Ok, I understand now ! Will do something like this, thanks

File: pollen-vision.md (hunk 1 above)
Comment (apirrone) [reply pollen-vision.md]: Ok we removed the link to google drive and decided it's actually better to only link to our github repo so this kind of information is centralized there

**Review by osanseviero — APPROVED** [review]
LGTM once the pending TODOs are solved!

**Comment by apirrone** [comment]
> LGTM once the pending TODOs are solved!

Thanks @osanseviero, I am just waiting for [this](https://huggingface.co/datasets/huggingface/documentation-images/discussions/295) to be merged to include the missing image links

**Comment by pcuenca** [comment]
@apirrone I just merged it (I realized after the fact that they should have gone inside a subfolder of the `blog` folder, but it's ok; if you need to upload more or update them then we can move them).

File: pollen-vision.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+We focused on selecting [zero-shot models](https://huggingface.co/tasks/zero-shot-object-detection), eliminating the need for any training, and making these tools instantly usable right out of the box.
+
+Our initial release is focused on 3D object detection—laying the groundwork for tasks like robotic grasping by providing a reliable estimate of objects' spatial coordinates. Currently limited to positioning within a 3D space (not extending to full 6D pose estimation), this functionality establishes a solid foundation for basic robotic manipulation tasks.
+
+### The Core Models of Pollen-Vision
+
+The library encapsulates several key models. We want the models we use to be zero-shot and versatile, allowing a wide range of detectable objects without re-training. The models also have to be “real-time capable”, meaning they should run at least a few fps on a consumer GPU. The first models we chose are:
```
Comment (pcuenca) [inline pollen-vision.md:30]:
```suggestion
The library encapsulates several key models. We want the models we use to be zero-shot and versatile, allowing a wide range of detectable objects without re-training. The models also have to be “real-time capable”, meaning they should run at least at a few fps on a consumer GPU. The first models we chose are:
```

File: pollen-vision.md [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+We're excited to share the initial launch of our open-source `pollen-vision` library, a first step towards empowering our robots with the autonomy to grasp unknown objects. **This library is a carefully curated collection of vision models chosen for their direct applicability to robotics.** `Pollen-vision` is designed for ease of installation and use, composed of independent modules that can be combined to create a 3D object detection pipeline, getting the position of the objects in 3D space (x, y, z). 
+
+We focused on selecting [zero-shot models](https://huggingface.co/tasks/zero-shot-object-detection), eliminating the need for any training, and making these tools instantly usable right out of the box.
+
+Our initial release is focused on 3D object detection—laying the groundwork for tasks like robotic grasping by providing a reliable estimate of objects' spatial coordinates. Currently limited to positioning within a 3D space (not extending to full 6D pose estimation), this functionality establishes a solid foundation for basic robotic manipulation tasks.
+
+### The Core Models of Pollen-Vision
```
Comment (pcuenca) [inline pollen-vision.md:28]:
```suggestion
## The Core Models of Pollen-Vision
```

File: pollen-vision.md [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+The library encapsulates several key models. We want the models we use to be zero-shot and versatile, allowing a wide range of detectable objects without re-training. The models also have to be “real-time capable”, meaning they should run at least a few fps on a consumer GPU. The first models we chose are:
+
+- [OWL-VIT](https://huggingface.co/docs/transformers/model_doc/owlvit) (Open World Localization - Vision Transformer, By Google Research): This model performs text-conditioned zero-shot 2D object localization in RGB images. It outputs bounding boxes (like YOLO)
+- [Mobile Sam](https://github.com/ChaoningZhang/MobileSAM): A lightweight version of the Segment Anything Model (SAM) by Meta AI. SAM is a zero-shot image segmentation model. It can be prompted with bounding boxes or points. 
+- [RAM](https://github.com/xinyu1205/recognize-anything) (Recognize Anything Model by OPPO Research Institute): Designed for zero-shot image tagging, RAM can determine the presence of an object in an image based on textual descriptions, laying the groundwork for further analysis. 
+
+### Get started in very few lines of code!
```
Comment (pcuenca) [inline pollen-vision.md:36]:
```suggestion
## Get started in very few lines of code!
```

File: pollen-vision.md [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+sam = MobileSamWrapper()
+annotator = Annotator()
+
+im = ...
+predictions = owl.infer(im, ["paper cups"])  # zero-shot object detection
+bboxes = get_bboxes(predictions)
+
+masks = sam.infer(im, bboxes=bboxes)  # zero-shot object segmentation
+annotated_im = annotator.annotate(im, predictions, masks=masks)
+```
```
Comment (pcuenca) [inline pollen-vision.md:55]: It'd be nice if we could show an example of the final result here :)

File: pollen-vision.md [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+
+
+OWL-VIT’s inference time depends on the number of prompts provided. On a Laptop with a RTX 3070 GPU:
```
Comment (pcuenca) [inline pollen-vision.md:58]:
```suggestion
OWL-VIT’s inference time depends on the number of prompts provided (i.e., the number of objects to detect). On a Laptop with a RTX 3070 GPU: 
```

File: pollen-vision.md [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+```
+1 prompt   : ~75ms  per frame
+2 prompts  : ~130ms per frame
+3 prompts  : ~180ms per frame
+4 prompts  : ~240ms per frame
+5 prompts  : ~330ms per frame
+10 prompts : ~650ms per frame
+```
+
+So it is interesting, performance-wise, to only prompt OWL-VIT with objects that we know are in the image. That’s where RAM is useful, as it is fast and provides exactly this information.
```
Comment (pcuenca) [inline pollen-vision.md:69]: But it's not used in the example, is it?

File: pollen-vision.md [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+
+### A robotics use case: grasping unknown objects in unconstrained environments
```
Comment (pcuenca) [inline pollen-vision.md:71]:
```suggestion
## A robotics use case: grasping unknown objects in unconstrained environments
```

**Review by pcuenca — APPROVED** [review]
Very cool! 🔥

**Comment by apirrone** [comment]
> @apirrone I just merged it (I realized after the fact that they should have gone inside a subfolder of the `blog` folder, but it's ok; if you need to upload more or update them then we can move them).

Ah sorry about that, Thanks!

File: pollen-vision.md (hunk 8 above)
Comment (apirrone) [reply pollen-vision.md:55]: Ah! There actually was an example there originally, it disappeared somehow :) 

I would need to add a new image in the documentation repo, can we move the images in a subfolder as you suggested here https://github.com/huggingface/blog/pull/1927#issuecomment-2014636869 ?

File: _blog.yml [hunk 12]
```diff
@@ -3676,6 +3675,17 @@
...
+  title: "Pollen-Vision: Unified interface for Zero-Shot vision models in robotics"
+  author: apirrone
+  guest: true
+  thumbnail: /blog/assets/pollen-vision/thumbnail.jpg
+  date: March 21, 2024
+  tags:
+    - robotics
+    - vision
+    - object-detection
+
```
[the remark this answers is not in the document]
Comment (apirrone) [reply _blog.yml]: Sorry about that !

File: pollen-vision.md (hunk 10 above)
Comment (apirrone) [reply pollen-vision.md:69]: It's not indeed. We wanted to keep the code snipped as small as possible. We will provide an example of the full pipeline in the repo

**Comment by apirrone** [comment]
@omarespejel @pcuenca 

There are many ways to embed .mp4 videos in markdown. Which one should I choose for the huggingface blog post format ?

Thanks !

File: pollen-vision.md (hunk 8 above)
Comment (pcuenca) [reply pollen-vision.md:55]: Yes, feel free to open a PR :)

**Comment by apirrone** [comment]
> > There are many ways to embed .mp4 videos in markdown. Which one should I choose for the huggingface blog post format ?
> 
> We use html, for example: https://github.com/huggingface/blog/blob/main/assisted-generation.md?plain=1#L20-L26

Perfect, thanks !

File: pollen-vision.md (hunk 8 above)
Comment (apirrone) [reply pollen-vision.md:57]: It's probably not the right way to do this, but I created a temporary empty file in `blog/pollen-vision/` in order to create the folder. Once it's merged I can upload all the files inside this folder and remove the tmp file :)

**Review by pcuenca — APPROVED** [review]

**Comment by apirrone** [comment]
All is good on our side, this can be merged if you think it's ready @pcuenca @osanseviero !

Thank you both for your thorough reviews !

**Comment by pcuenca** [comment]
Thank you @apirrone!

**Pull request merged**